jena-tdb1/src/main/java/org/apache/jena/tdb/sys/DatasetControlMRSW.java

```diff
@@ -126,6 +127,13 @@ public T next() {
             }
         }
 
+        @Override
+        public void forEachRemaining(Consumer<? super T> action) {
+            iter.forEachRemaining(action);
+            close();
+            checkIterConcurrentModification();
```

Comment:
Should this be at the start of the method?

Reply:
Sorry, I intruduced  a bug here, because `close `"disables" `checkIterConcurrentModification` in this class. So `close` have to be done after `checkIterConcurrentModification`.

I took a look into the Java `ArrayList` and `HashMap` doc and implementatons for my decision to put the check at the end.
The doc header of `HashMap` says;

> Note that the fail-fast behavior of an iterator cannot be guaranteed
>  as it is, generally speaking, impossible to make any hard guarantees in the
>  presence of unsynchronized concurrent modification.  Fail-fast iterators
>  throw `ConcurrentModificationException` on a best-effort basis.
>  Therefore, it would be wrong to write a program that depended on this
>  exception for its correctness: _the fail-fast behavior of iterators_
>  _should be used only to detect bugs_.

My conlusions:
`Iterator#hasNext` --> should not check for concurrent modifications
'Iterator#next' --> should check for concurrent modifications
'Iterator#forEachRemaining' -->  should usualy be checked at the end because this is the point where it is most useful to detect errors. Checking only at the beginning is almost useless. Checking at each iteration is expensive, but can be useful if it avoids index errors at the same time.

Here is one example for checking at the end: `java.util.HashMap.KeySet#forEach`.

Reply:
Thank you for the analysis.

A CCME is only "advice" - and then more for the same thread/nested changes case than multithread concurrent access.

Reply:
Thanks. 
If it is more for the same thread/nested changes, should I put it at the beginning of the implementation of "forEachRemaining" in "DatasetControlMRSW"?

Reply:
In the `forEachRemaining`, when not doing it fine grained in `next`, either start or finish is fine. It is not going to trap the moment the nested change happens, nor if the `forEachRemaining` is beign nested but that's OK.

Ready to merge?

Reply:
Yes, I would be happy if you would merge this PR.